fix: prepare for far classes

erights opened the pull request:
In order to support https://github.com/Agoric/agoric-sdk/pull/5970 , we need to allow at least one level of remotable inheritance. Currently supported in agoric-sdk as a patch. TODO after this is merged and agoric-sdk updated to depend on it, remove the patch.

erights commented:
ping

gibson042 reviewed (changes requested):
I'm requesting confirmation on the recursion depth question.

Also, this review uncovered behavior that smells like a bug: nothing checks if a prototype is hardened, so it can change underneath `passStyleOf`.
```js
import "@endo/init";
import { Far, passStyleOf, PASS_STYLE } from "@endo/marshal";

const nonEnumerable = { configurable: true, writable: true };
const mercurialProto = new Proxy(
  Object.defineProperties({}, {
    [PASS_STYLE]: { ...nonEnumerable, value: "remotable" },
    [Symbol.toStringTag]: { ...nonEnumerable, value: "Remotable" },
  }),
  {
    getPrototypeOf(obj) {
      // Self-mutate after returning the original prototype one time
      // (to checkRemotableProtoOf).
      if (obj[PASS_STYLE] !== "error") {
        obj[PASS_STYLE] = "error";
        return Object.prototype;
      }
      return Error.prototype;
    },
  }
);

const makeInput = () => Object.freeze({ __proto__: mercurialProto });
const input1 = makeInput();
const input2 = makeInput();

console.log("# passStyleOf(input1)");
console.log(passStyleOf(input1)); // => "remotable"

/* same because of passStyleMemo WeakMap */
console.log(`# passStyleOf(input1) again (cached "Purely for performance")`);
console.log(passStyleOf(input1)); // => "remotable"

/* different because of changes in the prototype */
console.log("# passStyleOf(input2)");
console.log(passStyleOf(input2)); // => Error (Errors must inherit from an error class .prototype)
```

erights commented:
> Also, this review uncovered behavior that smells like a bug: nothing checks if a prototype is hardened, so it can change underneath passStyleOf.

Hi @gibson042 , when you tried this experiment, was that on a system that was pre or post #1250 ? Your attack seems much like https://github.com/Agoric/agoric-sdk/issues/4333 , which I believe I fixed in #1250, and which this PR relies on. Note that I fixed 4333 only in #1250 in endo. I did *not* fix the patch in agoric-sdk, so it you're testing that, your attack should still work.

In any case, worried that #1250 didn't really fix it, I tried your test case, which I will add to this PR. I'm pleased to report that your first call to `passStyleOf` threw `'A tagRecord must be frozen: "[undefined: undefined]"'`.

Nevertheless, I will carefully review the code and see if any variation of this attack seems possible. PTAL and see if you can still find any way to attack the system.

erights commented on packages/marshal/test/test-passStyleOf.js:
```diff
@@ -129,7 +128,5 @@ test('passStyleOf testing remotables', t => {
   const farObj6 = harden({
     __proto__: farObjProto6,
```
I don't have strong feeling about this, so am happy to switch if you do.

The reasons I like it better:
   * It is syntax, and so its meaning is invariant of any bindings in the local lexical scope.
   * The properties of the object, which I admit are absent here, can appear using normal object-literal property syntax, rather than descriptor syntax, which is more pleasant when that it all you need.

I'll leave this unresolved until I hear back.

erights replied:
Took your suggested helper. Definite improvement. Thanks. Done.

erights commented on packages/marshal/src/helpers/remotable.js:
```diff
@@ -77,6 +77,18 @@ const checkRemotableProtoOf = (original, check = x => x) => {
    *        }}
    */
   const proto = getPrototypeOf(original);
+  const protoProto = getPrototypeOf(proto);
```
I do intend arbitrary-depth recursion. So I accepted this suggestion but changed the comment. I'll explain elsewhere why I intend arbitrary-depth recursion. But this one is done. Resolving.

erights commented on packages/marshal/src/helpers/remotable.js:
```diff
@@ -77,6 +77,18 @@ const checkRemotableProtoOf = (original, check = x => x) => {
+  ) {
+    return (
+      // eslint-disable-next-line no-use-before-define
+      RemotableHelper.canBeValid(proto, check) && checkRemotable(proto, check)
```
I do intend the arbitrary depth recursion, even though agoric-sdk currently uses only one or two. The reason is that we're explaining FarClasses as class-like, making things more intuitive for those used to programming with classes. It would be natural for them to expect to be able to define FarClasses that inherit from other FarClasses. The new internal representation in this PR could naturally support that, and we may want to provide that.

Having experienced the coordination difficulties of agoric-sdk needing endo to expand its remotable recognition rules (the need to patch. the breakage of CI because of something that doesn't see the patch), I thought I'd get ahead of the game and prepare to accept that deeper recursion. AFAICT, there's no reason not to, and it keep options open both for us, and for others who want to build new representations on top of endo.

Resolving.

erights commented on packages/marshal/test/test-passStyleOf.js:
```diff
@@ -129,7 +128,5 @@ test('passStyleOf testing remotables', t => {
-  t.throws(() => passStyleOf(farObj6), {
-    message: /"\[Symbol\(passStyle\)\]" property expected: "\[Alleged: manually constructed\]"/,
-  });
+  t.is(passStyleOf(farObj6), 'remotable');
```
Whew, done! PTAL

erights commented on packages/marshal/test/test-passStyleOf.js:
```diff
@@ -129,7 +131,155 @@ test('passStyleOf testing remotables', t => {
+    message: 'For now, remotables cannot inherit from anything unusual, in {}',
+  });
+
+  const tagRecordA = Object.create(Object.prototype, {
+    __proto__: null,
+    [PASS_STYLE]: { value: 'remotable' },
+    [Symbol.toStringTag]: { value: 'Alleged: null grandproto is fine' },
+  });
+  const farObjProtoA = harden({
+    __proto__: tagRecordA,
+  });
+  const farObjA = harden({
+    __proto__: farObjProtoA,
+  });
+  t.is(passStyleOf(farObjA), 'remotable');
```
Hey thanks! Your fixing my misplaced null caused the test to fail because the code being tested had a bug. My incorrect test code masked that bug.

erights replied:
Done, with three variations.
